### PR TITLE
Introduce boolean literal freshness

### LINKED ISSUE #25474: Literal type `false` unexpectedly widens into type `boolean`
**TypeScript Version:**  3.0.0-dev.20180705

**Search Terms:** false/true loosened/expanded/loses specificity into boolean

**Code**

```ts
type Foo = FooBase | FooArray;
type FooBase = string | false;
type FooArray = FooBase[];

declare let foo1: Foo;
declare let foo2: Foo;
foo1 = [...Array.isArray(foo2) ? foo2 : [foo2]];
```

**Expected behavior:**
No error.. In the else clause, `[foo2]`, it retains `string | false` so `Array<string | false>`. Logically since everything is taking place in the typings layer, `foo2` should still be strongly bounded to `false` and will never be `true`.

**Actual behavior:**
Errors. In the else clause, `[foo2]`, it expanded/loosened to `string | boolean` so `Array<string | boolean>`.

```
tmp.ts:7:1 - error TS2322: Type '(string | boolean)[]' is not assignable to type 'Foo'.
  Type '(string | boolean)[]' is not assignable to type 'FooBase[]'.
    Type 'string | boolean' is not assignable to type 'FooBase'.
      Type 'true' is not assignable to type 'FooBase'.

7 foo1 = [...Array.isArray(foo2) ? foo2 : [foo2]];
  ~~~~
```

**Playground Link:** [Link](https://www.typescriptlang.org/play/#src=type%20Foo%20%3D%20FooBase%20%7C%20FooArray%3B%0D%0Atype%20FooBase%20%3D%20string%20%7C%20false%3B%0D%0Atype%20FooArray%20%3D%20FooBase%5B%5D%3B%0D%0A%0D%0Adeclare%20let%20foo1%3A%20Foo%3B%0D%0Adeclare%20let%20foo2%3A%20Foo%3B%0D%0Afoo1%20%3D%20%5B...Array.isArray(foo2)%20%3F%20foo2%20%3A%20%5Bfoo2%5D%5D%3B%0D%0A)

**Related Issues:**

### PR DESCRIPTION
**Pull request by weswigham**
The logical follwup to adding enum freshness last week, as we mentioned at the time. I hadn't realized that we _had_ had issues reported which required it to be fixed, but we had, so here's the freshness implementation for booleans. It's rather straightforward since all the variants can be constructed up front. The old `trueType`/`falseType` map to the new fresh variant. The change also brings a noticeable improvement to the accuracy of CFA when multiple boolean-typed declarations are involved, as they no longer always widen irrecoverably to `boolean`.

The only literal types which do not have freshness of any kind now are, I believe, tuples.

Fixes #25474

### CODE REVIEW

**Comment by weswigham** [comment]
@typescript-bot test this

**Comment by weswigham** [comment]
@typescript-bot test this again

File: src/compiler/types.ts [hunk 1]
```diff
@@ -3731,7 +3731,7 @@ namespace ts {
         Unit = Literal | UniqueESSymbol | Nullable,
         StringOrNumberLiteral = StringLiteral | NumberLiteral,
         /* @internal */
-        StringOrNumberLiteralOrUnique = StringOrNumberLiteral | UniqueESSymbol,
+        StringOrNumberLiteralOrUnique = StringLiteral | NumberLiteral | UniqueESSymbol,
```
Comment (sandersn) [inline src/compiler/types.ts]: this change doesn't seem related (and I'm not sure it's an improvement?)

Comment (weswigham) [reply src/compiler/types.ts]: Eh, I changed ti because I actually removed `StringOrNumberLiteral` at first because it was practically unused within our code, but then added ti back to avoid an API break.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -14092,7 +14106,10 @@ namespace ts {
                 if (assignedType.flags & TypeFlags.Never) {
                     return assignedType;
                 }
-                const reducedType = filterType(declaredType, t => typeMaybeAssignableTo(assignedType, t));
+                let reducedType = filterType(declaredType, t => typeMaybeAssignableTo(assignedType, t));
+                if (assignedType.flags & (TypeFlags.FreshLiteral | TypeFlags.Literal)) {
+                    reducedType = mapType(reducedType, getFreshTypeOfLiteralType); // Ensure that if the assignment is a fresh type, that we narrow to fresh types
```
Comment (sandersn) [inline src/compiler/checker.ts]: why didn't we have to do this before, for string and number literal types?

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -405,9 +405,22 @@ namespace ts {
...
+        trueType.regularType = regularTrueType;
+        regularTrueType.freshType = trueType;
+        falseType.regularType = regularFalseType;
+        regularFalseType.freshType = falseType;
+        const booleanType = createBooleanType([regularFalseType, regularTrueType]);
+        // Also mark all combinations of fresh/regular booleans as "Boolean" so they print as `boolean` instead of `true | false`
+        // (The union is cached, so simply doing the marking here is sufficient)
+        createBooleanType([regularFalseType, trueType]);
+        createBooleanType([falseType, regularTrueType]);
+        createBooleanType([falseType, trueType]);
```
Comment (sandersn) [inline src/compiler/checker.ts]: can we ever get a union of `falseType | regularFalseType`? Note that it doesn't matter how we print that, I'm just curious whether it can happen.

File: src/compiler/checker.ts (hunk 2 above)
Comment (weswigham) [reply src/compiler/checker.ts]: _We probably should have been_ but as it turns out people narrow `boolean` made via `!!expr || !expr` a ton more than narrowing `x === "a" | x === "b"`, so it's a lot more obvious an issue with booleans.

File: src/compiler/checker.ts (hunk 3 above)
Comment (weswigham) [reply src/compiler/checker.ts]: Uhmmm, I don't think so, just like how we can't have ``"word" | "word"(fresh edition)`, sicne both have the same type "identity". Specifically, `removeRedundantLiteralTypes` handles removing redundant regular types when the fresh one is also present.

File: src/compiler/checker.ts (hunk 2 above)
Comment (weswigham) [reply src/compiler/checker.ts]: And as I said in person, this is going to break one RWC project along the lines of this:
```ts
type RuleSeverity = "warning" | "error" | "off";
interface IOptions {
    ruleSeverity: RuleSeverity;
}

function parseRuleOptions(): Partial<IOptions> {
    let defaultRuleSeverity: RuleSeverity = "error";
    let ruleSeverity = defaultRuleSeverity;

    return {
        ruleSeverity,
    };
}
```
previously this worked because `defaultRuleSeverity` was narrowed to the nonfresh `"error"` when it was used to initialize `ruleSeverity`, whereas now we choose the fresh `"error"` which widens (and triggers an error in the return type below). So this is a (probably small) break.

**Comment by aleksey-bykov** [comment]
what is freshness please?

**Comment by sandersn** [comment]
@aleksey-bykov I wrote [a novel-length treatment](
https://sandersn.github.io/manual/Widening-and-Narrowing-in-Typescript.html) of the different ways the compiler converts types to related types. Note that it *still* doesn’t cover the base constraint type, which is related to the apparent type. But it does cover freshness pretty well.

File: src/compiler/checker.ts (hunk 2 above)
Comment (weswigham) [reply src/compiler/checker.ts]: FWIW the individual unit type only behaves differently because it skips the narrowing codepath. Had I realized as much I probably would've altered the unit type behavior, too. In any case, removing this bit if we don't like it is easy; but it's going to make boolean literals a tad more breaky.
